Add best practices for tooling (#551)
exercism/docs: building/tooling/best-practices.md

As each test run happens in a fresh container, this means that this compilation was done _in every single test run_!
To circumvent this, the [Haskell test runner's Dockerfile](https://github.com/exercism/haskell-test-runner/blob/5264c460054649fc672c3d5932c2f3cb082e2405/Dockerfile) has the following two commands:

```dockerfile
COPY pre-compiled/ .
RUN stack build --resolver lts-20.18 --no-terminal --test --no-run-tests
```

First, the `pre-compiled` directory is copied into the image.
This directory is setup as a sort of fake exercise and depends on the same base libraries that the actual exercise depend on.
Then we run the tests on that directory, which is similar to how tests are run for an actual exercise.
Running the tests will result in the base being compiled, but the difference is that this happens at _build time_.
The resulting Docker image will thus have its base libraries already compiled, which means that no longer has to happen at _run time_, resulting in (much) faster execution times.

#### Example: pre-compile binaries

Some languages allow code to be compiled ahead-of-time or just-in-time.

> This is a build time vs. run time tradeoff, and again, we favor build time execution for performance reasons

Reviewer: ```suggestion
This is a build time vs. run time tradeoff, and again, we favor build time execution for performance reasons.
```

> The [C# test runner's Dockerfile](https://github.com/exercism/csharp-test-runner/blob/b54122ef76cbf86eff0691daa33c8e50bc83979f/Dockerfile) uses this approach, where the test runner is compiled t a binary ahead-of-time (at build time) instead of just-in-time compiling the code (at run time).

Reviewer: ```suggestion
The [C# test runner's Dockerfile](https://github.com/exercism/csharp-test-runner/blob/b54122ef76cbf86eff0691daa33c8e50bc83979f/Dockerfile) uses this approach, where the test runner is compiled to a binary ahead-of-time (at build time) instead of just-in-time compiling the code (at run time).
```

> This mean that there is less work to do a run time, which should help increase performance.

Reviewer: ```suggestion
This means that there is less work to do at run-time, which should help increase performance.
```

## Size

> You should try to reduce the image's size, which means that it'll be:

Reviewer: ```suggestion
You should try to reduce the image's size, which means that it'll:
```

> - Faster to deploy

Reviewer: ```suggestion
- Be faster to deploy
```

> - Reduces costs for us

Reviewer: ```suggestion
- Reduce costs for us
```

> - Marginally improves startup time of each container

Reviewer: ```suggestion
- Improve startup time of each container
```

```
REPOSITORY   TAG            SIZE
node         20.16.0        1.09GB
node         20.16.0-slim   219MB
```

The reason "slim" variants are smaller is that they'll have less features.
Your image might not need the additional features, and if not, consider using the "slim" variant.

### Removing unneeded bits

An obvious, but great, way to reduce the size of your image is to remove anything you don't need.
These can include things like:

- Source files that are no longer needed after building a binary from them
- Files targeting different architectures from the Docker image
- Documentation

#### Remove package manager files

Most Docker images need to install additional packages, which is usually done via a package manager.
These packages must be installed at _build time_ (as no internet connection is available at _run time_).
Therefore, any package manager caching/bookkeeping files should be removed after installing the additional packages.

##### apk

Distributions that uses the `apk` package manager (such as Alpine) should use the `--no-cache` flag when using `apk add` to install packages:

```dockerfile
RUN apk add --no-cache curl
```

##### apt-get/apt

> Distributions that uses the `apt-get`/`apk` package manager (such as Ubuntu) should run the `apt-get autoremove -y` and `rm -rf /var/lib/apt/lists/*` commands _after_ installing the packages:

Reviewer: ```suggestion
Distributions that use the `apt-get`/`apk` package manager (such as Ubuntu) should run the `apt-get autoremove -y` and `rm -rf /var/lib/apt/lists/*` commands _after_ installing the packages:
```

Multi-stage builds are particularly useful when your Dockerfile requires packages to be installed that are _only_ needed at build time.
In this situation, the general structure of your Dockerfile looks like this:

1. Define a new stage (we'll call this the "build" stage).
   This stage will _only_ be used at build time.
2. Install the required additional packages (into the "build" stage).
3. Run the commands that require the additional packages (within the "build" stage).
4. Define a new stage (we'll call this the "runtime" stage).
   This stage will make up the resulting Docker image and executed at run time.
5. Copy the result(s) from the commands run in step 3 (in the "build" stage) into this stage (the "runtime" stage).

With this setup, the additional packages are _only_ installed in the "build" stage and _not_ in the "runtime" stage, which means that they won't end up in the Docker image that is produced.

#### Example: downloading files

The Fortran test runner requires `curl` to download some files.
However, its run time image does _not_ need `curl`, which makes this a perfect use case for a multi-stage build.

> First, it's [Dockerfile](https://github.com/exercism/fortran-test-runner/blob/783e228d8449143d2040e68b95128bb791833a27/Dockerfile) defines a stage (named "build") in which the `curl` package is installed.

Reviewer: ```suggestion
First, its [Dockerfile](https://github.com/exercism/fortran-test-runner/blob/783e228d8449143d2040e68b95128bb791833a27/Dockerfile) defines a stage (named "build") in which the `curl` package is installed.
```
